omod/src/main/java/org/openmrs/module/adminui/page/controller/myaccount/ChangeSecretQuestionPageController.java

```diff
@@ -9,98 +9,72 @@
-            InfoErrorMessageUtil.flashInfoMessage(request.getSession(), "adminui.account.secretQuestion.success");
-        } catch (Exception ex) {
-            request.getSession().setAttribute(UiCommonsConstants.SESSION_ATTRIBUTE_ERROR_MESSAGE, "registrationapp.save.fail");
-            return "myaccount/changeSecretQuestion";
-        }
-        return "myaccount/myAccount";
-    }
-
-    public class SecretQuestion {
-
-        private String password;
-        private String question;
-        private String answer;
-        private String confirmAnswer;
-
-        public SecretQuestion(String password, String question, String answer, String confirmAnswer) {
-            this.password = password;
-            this.question = question;
-            this.answer = answer;
-            this.confirmAnswer = confirmAnswer;
-        }
-
-        public String getPassword() {
-            return password;
-        }
-
-        public void setPassword(String password) {
-            this.password = password;
-        }
-
-        public String getQuestion() {
-            return question;
-        }
-
-        public void setQuestion(String question) {
-            this.question = question;
-        }
-
-        public String getAnswer() {
-            return answer;
-        }
-
-        public void setAnswer(String answer) {
-            this.answer = answer;
-        }
-
-        public String getConfirmAnswer() {
-            return confirmAnswer;
-        }
-
-        public void setConfirmAnswer(String confirmAnswer) {
-            this.confirmAnswer = confirmAnswer;
-        }
-
-        @Override
-        public String toString() {
-            return "Question=" + question + ", Answer=" + answer;
-        }
-    }
+	
+	public void get(PageModel model) {
+		String secretQuestion = fetchExistingSecretQuestionOrEmpty();
+		model.put("secretQuestion", secretQuestion);
+	}
+	
+	public String post(PageModel model, @RequestParam("password") String password,
+	                   @RequestParam("question") String question, @RequestParam("answer") String answer,
+	                   @RequestParam("confirmAnswer") String confirmAnswer,
+	                   @SpringBean("userService") UserService userService, HttpServletRequest request) {
+		if (!answer.equals(confirmAnswer)) {
+			model.put("secretQuestion", fetchExistingSecretQuestionOrEmpty());
+			request.getSession().setAttribute(UiCommonsConstants.SESSION_ATTRIBUTE_ERROR_MESSAGE,
+			    Context.getMessageSourceService().getMessage("adminui.account.answerAndConfirmAnswer.doesNotMatch"));
+			
+			return "myaccount/changeSecretQuestion";
+		} else {
+			return changeSecretQuestion(password, question, answer, userService, request, model);
+		}
+	}
+	
+	private String changeSecretQuestion(String password, String question, String answer, UserService userService,
+	                                    HttpServletRequest request, PageModel model) {
+		try {
```

Comment:
The ticket description says: When the form is submitted and the user has provided any of the blank fields or the question has been changed, all fields should have been provided, if not require them. If all fields are provided and valid, save changes otherwise don't do anything.